clickhouse-admin/src/context.rs

```diff
@@ -3,84 +3,555 @@
+    fn fmri(self) -> String {
+        match self {
+            NodeSettings::Replica { fmri, .. } => fmri,
+            NodeSettings::Keeper { fmri, .. } => fmri,
+        }
+    }
+
+    fn generation(&self) -> Generation {
+        match self {
+            NodeSettings::Replica { settings, .. } => settings.generation(),
+            NodeSettings::Keeper { settings, .. } => settings.generation(),
+        }
+    }
+}
+
+pub fn generate_config_and_enable_svc(
+    generation_tx: watch::Sender<Option<Generation>>,
+    clickward: Clickward,
+    node_settings: NodeSettings,
+) -> Result<GenerateConfigResult, HttpError> {
+    let incoming_generation = node_settings.generation();
+    let generation_rx = generation_tx.subscribe();
+    let current_generation = *generation_rx.borrow();
+
+    // If the incoming generation number is lower, then we have a problem.
+    // We should return an error instead of silently skipping the configuration
+    // file generation.
+    if let Some(current) = current_generation {
+        if current > incoming_generation {
+            return Err(HttpError::for_client_error(
+                Some(String::from("Conflict")),
+                StatusCode::CONFLICT,
+                format!(
+                    "current generation '{}' is greater than incoming generation '{}'",
+                    current,
+                    incoming_generation,
+                )
+            ));
+        }
+    };
+
+    let output = match node_settings {
+        NodeSettings::Replica { ref settings, .. } => {
+            GenerateConfigResult::Replica(
+                clickward.generate_server_config(settings)?,
+            )
+        }
+        NodeSettings::Keeper { ref settings, .. } => {
+            GenerateConfigResult::Keeper(
+                clickward.generate_keeper_config(settings)?,
+            )
+        }
+    };
+
+    // We want to update the generation number only if the config file has been
+    // generated successfully.
+    generation_tx.send(Some(incoming_generation)).map_err(|e| {
+        HttpError::for_internal_error(format!("failure to send request: {e}"))
+    })?;
+
+    // Once we have generated the client we can safely enable the clickhouse_server service
+    Svcadm::enable_service(node_settings.fmri())?;
+
+    Ok(output)
+}
+
+pub async fn init_db(
+    clickhouse_address: SocketAddrV6,
+    log: Logger,
+    replicated: bool,
+) -> Result<(), HttpError> {
+    // We initialise the Oximeter client here instead of keeping it as part of the context as
+    // this client is not cloneable, copyable by design.
+    let client = OximeterClient::new(clickhouse_address.into(), &log);
+
+    // Initialize the database only if it was not previously initialized.
+    // TODO: Migrate schema to newer version without wiping data.
+    let version = client.read_latest_version().await.map_err(|e| {
+        HttpError::for_internal_error(format!(
+            "can't read ClickHouse version: {e}",
```

Comment:
Including the error here is good, but as written `{e}` will only include the top-most error and not the full error chain. Can we add a dependency on `slog-error-chain` and use it to get the full chain of errors?

```suggestion
            "can't read ClickHouse version: {}", InlineErrorChain::new(e),
```